src/neo4clj/query_builder.clj

```diff
@@ -62,81 +62,139 @@
+  node entries or directly on the given node if not in the list.
+
+  The optional flag as-lookup? dictates wheter the query should be a
+  lookup or a raw node reference.
+
+  The optional flag unknown-only? returns nil for already known
+  ref-ids"
+  ([known-ref-ids node-entries node]
+   (node-reference known-ref-ids node-entries node true false))
+  ([known-ref-ids node-entries node as-lookup?]
+   (node-reference known-ref-ids node-entries node as-lookup? false))
+  ([known-ref-ids node-entries node as-lookup? unknown-only?]
+   (let [ref-id (node-ref-id node)]
+     (if-let [known-ref? (known-ref-ids ref-id)]
+       (when (not unknown-only?)
+         [(str "(" ref-id ")") nil])
+       ((if as-lookup? cypher/lookup-node #(vector (cypher/node %) nil)) (or (node-entries ref-id) node))))))
+
+(defn- generate-relation-queries
+  "Takes a set of known reference id's, a map of node entries to
+  lookup and a list of relationships to create queries for.
+
+  If a nodes reference id is already known, the node is reference by
+  it, else a lookup query is inserted instead."
+  [known-ref-ids node-entries rels as-lookup?]
+  (loop [remaining-rels rels
+         known-refs known-ref-ids
+         rel-queries []
+         where-parts []]
+      (if (empty? remaining-rels)
+        [rel-queries where-parts known-refs]
+        (let [{:keys [from to] :as rel} (first remaining-rels)
+              [from-query from-where] (node-reference known-refs node-entries from as-lookup?)
+              [to-query to-where] (node-reference known-refs node-entries to as-lookup?)]
+          (recur (rest remaining-rels)
+                 (conj known-refs (node-ref-id from) (node-ref-id to))
+                 (conj rel-queries (cypher/relationship from-query to-query rel))
+                 (concat where-parts (remove nil? [from-where to-where])))))))
 
 (defn create-graph-query
   "Takes a graph representation and creates the nodes and relationship
   defined and returns any aliases specified in the representation"
   [{:keys [lookups nodes rels returns]}]
   (str/join " " (concat
                  (map #(lookup-node % false) lookups)
-                 (map #(create-node-query % false) nodes)
-                 (map create-graph-rel-query rels)
+                 [(str "CREATE " (str/join " CREATE " (first (generate-relation-queries
+                                                              (set (map :ref-id lookups))
+                                                              (reduce #(assoc %1 (:ref-id %2) %2) {} nodes)
+                                                              rels
+                                                              false))))]
                  (when-not (empty? returns)
-                   (vector (str "RETURN " (str/join "," (get-return-ref-ids returns))))))))
+                   (vector (str "RETURN " (str/join ", " (get-return-ref-ids returns))))))))
+
+(defn- lookup-unmatched-nodes
```

Comment:
I'm having a hard time giving feedback on `node-reference`, `generate-relation-queries` and `lookup-unmatched-nodes` because they are complex. One thing that might be able to help would be a test case but then they cant be "private".

Reply:
I think it's better to ensure they are not called from outside as they are very specific helper functions.

Reply:
This is something we've discussed before. My take would be to have two namespaces:
- One with the "public" library API exposed to world.
- Another one with the "helper functions".

This way you are relying on how Java deals with stuff. This pattern will not work in [ClojureScript (see `:private` under Special Forms)][1].

[1]: https://clojurescript.org/about/differences#_special_forms